cluster/node-pools/worker-karpenter/provisioners.yaml

```diff
@@ -0,0 +1,121 @@
+    - key: karpenter.sh/capacity-type
+      operator: In
+      values:
+        - "spot"
+        - "on-demand"
+    - key: "kubernetes.io/arch"
+      operator: In
+      values: ["arm64", "amd64"]
+    - key: "topology.kubernetes.io/zone"
+      operator: In
+      values:
+{{- $azs := .Values.availability_zones }}
+{{- if index .NodePool.ConfigItems "availability_zones"}}
+        {{    $azs = split .NodePool.ConfigItems.availability_zones "," }}
+{{- end}}
+{{- range $az :=  $azs }}
+      - "{{ $az }}"
+{{- end }}
+{{- $taints := .NodePool.Taints}}
+{{- if $taints }}
+  taints:
+  {{- range $taints }}
+    {{- $taint := . }}
+      - key: {{$taint.Key}}
+        effect: {{$taint.Effect}}
+        {{- if $taint.Value }}
+        value: {{$taint.Value }}
+        {{- end }}
+  {{- end}}
+{{- end}}
+
+  startupTaints:
+  - key: zalando.org/node-not-ready
+    effect: NoSchedule
+
+  labels:
+    # these labels are normally set by kubelet on start-up, but because
+    # karpenter already creates the node object ahead of the instance start, we
+    # need to set them here as well.
+    lifecycle-status: ready
+    node.kubernetes.io/node-pool: "{{ .NodePool.Name }}"
+    node.kubernetes.io/role: worker
+    node.kubernetes.io/profile: "{{ .NodePool.Profile }}"
+    cluster-lifecycle-controller.zalan.do/replacement-strategy: none
+{{- if index .NodePool.ConfigItems "labels"}}
+  {{- range split .NodePool.ConfigItems.labels ","}}
+    {{- $label := split . "="}}
+    {{index $label 0}}: {{index $label 1}}
+  {{- end}}
+{{- end}}
+
+  # ttlSecondsAfterEmpty: 30
+  providerRef:
+    name: {{ .NodePool.Name }}
+---
+apiVersion: karpenter.k8s.aws/v1alpha1
+kind: AWSNodeTemplate
+metadata:
+  name: {{ .NodePool.Name }}
+spec:
+  amiFamily: Custom
+  amiSelector:
+    aws-ids: "{{ .Cluster.ConfigItems.kuberuntu_image_v1_22_amd64  }},{{ .Cluster.ConfigItems.kuberuntu_image_v1_22_arm64 }}"
+  metadataOptions:
+    httpTokens: optional
+  subnetSelector:
+    Name: "dmz-eu-central-1*" # TODO: use a better tag + it has to have auto assign public IP enabled
+  securityGroupSelector:
+    karpenter.sh/discovery: "{{ .Cluster.ID }}/WorkerNodeSecurityGroup"
+  blockDeviceMappings:
+    - deviceName: /dev/sda1
+      ebs:
+        deleteOnTermination: {{ .NodePool.ConfigItems.ebs_root_volume_delete_on_termination}}
+        volumeSize: {{ .NodePool.ConfigItems.ebs_root_volume_size }}Gi
+        volumeType: gp3
+  userData: |
+{{.UserData | indent 4}}
+  tags:
+    InfrastructureComponent: "true"
+    # TODO: maybe use application label on dedicated node pools?
+    application: kubernetes
+    component: worker
+    environment: "{{ .Cluster.Environment }}"
+    Name: "{{ .NodePool.Name }} ({{ .Cluster.ID }})"
+    # TODO: are these tags still needed?
+    # Do we want a node pool tag for cost tracking reasons?
+#    node.kubernetes.io/role: worker
+#    node.kubernetes.io/node-pool: "{{ .NodePool.Name }}"
+#    kubernetes.io/cluster/{{ .Cluster.ID }}: "owned"
```

Comment:
Are these tags set from labels on the node or not set at all now?

Reply:
only the `kubernetes.io/cluster/` is set automatically, will add the other 2 tags

Reply:
karpenter also tag them with:
`karpenter.sh/provisioner-name`

but I guess we have some tooling around `node.kubernetes.io/node-pool`

Reply:
> but I guess we have some tooling around `node.kubernetes.io/node-pool`

I'm not 100% sure but I imagine we have some ZMON check/alerts that may depend on it and potentially some cost reporting could also. To avoid that we have to clean this up as part of karpenter it can be safer to just keep them and then address it later if we want to clean it up.

There are also things like the `dedicated` tag which can be useful e.g. in the case of skipper-ingress node pools to get their cost in AWS Cost Explorer. That would currently not work with Karpenter, but we also will not run skipper-ingress here right away so we can solve that later.

Reply:
I added the tags back, and also added the dedicated tag.